src/CoreServices/FSEvents.cs

```diff
@@ -259,12 +288,37 @@ static void FreeGCHandle (IntPtr gchandle)
 			}
 
 			var events = new FSEvent[numEvents];
-			var pathArray = new CFArray (eventPaths, false);
 
 			for (int i = 0; i < events.Length; i++) {
-				events[i].Flags = (FSEventStreamEventFlags)(uint)Marshal.ReadInt32 (eventFlags, i * 4);
-				events[i].Id = (uint)Marshal.ReadInt64 (eventIds, i * 8);
-				events[i].Path = CFString.FromHandle (pathArray.GetValue (i));
+				string? path = null;
+				ulong fileId = 0;
+
+				var eventDataHandle = CFArrayGetValueAtIndex (eventPaths, i);
+				var eventDataType = CFType.GetTypeID (eventDataHandle);
+
+				if (eventDataType == CFStringTypeID) {
+					path = CFString.FromHandle (eventDataHandle);
+				} else if (eventDataType == CFDictionaryTypeID) {
+					var pathHandle = CFDictionaryGetValue (
+						eventDataHandle,
+						kFSEventStreamEventExtendedDataPathKey.Handle);
+					if (pathHandle != IntPtr.Zero)
+						path = CFString.FromHandle (pathHandle);
+
+					var fileIdHandle = CFDictionaryGetValue (
+						eventDataHandle,
+						kFSEventStreamEventExtendedFileIDKey.Handle);
+					if (fileIdHandle != IntPtr.Zero)
+						CFNumberGetValue (fileIdHandle, 4 /*kCFNumberSInt64Type*/, out fileId);
+				}
+
+				events[i] = new FSEvent
+				{
+					Id = (ulong)Marshal.ReadInt64 (eventIds, i * 8),
```

Comment:
This looks like something that could be wrapped into a `ReadOnlySpan<ulong>()`. I think it's been done already in a similar fashion over at dotnet/runtime in FileSystemWatcher.OSX.cs

Reply:
Sure, but this code was just moved from a few lines above. This isn't new code, so I didn't go around changing things.

Reply:
Er, no, that would mean the underlying native alloc'd memory would have to stick around. It's releated once the callback returns.